provisioning/device/src/transport_state_machine.ts

```diff
@@ -0,0 +1,239 @@
+                  this._fsm.transition('waiting_to_poll', callback, registrationId, responseBody.operationId, pollingInterval);
+                  break;
+                }
+                default: {
+                  /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_016: [ If `TransportHandlers.registrationRequest` succeeds returns with an unknown status, `register` shall fail with a `SyntaxError` and pass the response body and the protocol-specific result to the `callback`. ] */
+                  /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_022: [ If `TransportHandlers.queryOperationStatus` succeeds with an unknown status, `register` shall fail with a `SyntaxError` and pass the response body and the protocol-specific result to the `callback`. ] */
+                  let err = new SyntaxError('status is ' + responseBody.status);
+                  (err as any).result = result;
+                  (err as any).responseBody = responseBody;
+                  this._fsm.transition('connected', callback, err);
+                  break;
+                }
+              }
+            } else {  // err == null && responseBody == null
+              let err = this._transport.getErrorResult(result);
+              (err as any).result = result;
+              (err as any).responseBody = responseBody;
+              this._fsm.transition('connected', callback, err, responseBody, result);
+            }
+          },
+          responseReceived: () => { throw new Error(); },
+          disconnect: (callback, err, responseBody) => this._fsm.transition('disconnecting', callback, err, responseBody),
+          /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_007: [ If `connect` is called while the transport is waiting between operation status queries, it shall do nothing and call `callback` immediately. ] */
+          connect: (callback) => callback(),
+          /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_024: [ If `register` is called while a different request is in progress, it shall fail with an `InvalidOperationError`. ] */
+          register: (callback) => callback(new errors.InvalidOperationError('another operation is in progress'))
+        },
+        waiting_to_poll: {
+          _onEnter: (callback, registrationId, operationId, pollingInterval) => {
+            debug('waiting for ' + pollingInterval + ' ms');
+            this._pollingTimer = setTimeout(() => {
+              this._fsm.transition('polling', callback, registrationId, operationId, pollingInterval);
+            }, pollingInterval);
+            this._pollingTimer.callback = callback;
+          },
+          disconnect: (callback, err, responseBody) => {
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_027: [ If a registration is in progress, `disconnect` shall cause that registration to fail with an `OperationCancelledError`. ] */
+            this._pollingTimer.callback(new errors.OperationCancelledError());
```

Comment:
>callback [](start = 31, length = 8)

same synchronous callback issue here - we're going to stay stuck in this state until the callback is executed, then transition to disconnecting. is this the desirable behavior? I feel like we should be transitioning, and once transitioned, call the necessary callbacks.

Reply:
one of the reason you want to be in the good state before calling anything is in case the callback triggers a retry.

---
In reply to: [147229585](https://github.com/Azure/azure-iot-sdk-node/pull/115#discussion_r147229585) [](ancestors = 147229585)